### PR TITLE
Justus/exercise03 (1306) 

### PR DESCRIPTION
**Pull request by Saiki77**

### CODE REVIEW

File: Exercises/Justus Exercises/Justus Exercises_03.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+# write a function that accepts two numbers, a lower bound and an upper bound.
+# the function should then return the count of all numbers that meet certain criteria:
+# - they are within the (left-inclusive and right-exclusive) bounds passed to the function
+# - there is at least one group of exactly two adjacent digits within the number which are the same (like 33 in 123345)
+# - digits only increase going from left to right
+
+def count_special_numbers(lower_bound, upper_bound):
+    def is_valid_number(num_str):
+        # Check if digits are increasing
+        if list(num_str) != sorted(num_str):
```
Comment (laraasphh) [inline Exercises/Justus Exercises/Justus Exercises_03.py:106]: Didn't know there was a sorted method to solve this too. I think it's cool and it looks clear

File: Exercises/Justus Exercises/Justus Exercises_03.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            return False
+        
+        # Check for at least one group of exactly two repeated digits
+        for i in range(len(num_str) - 1):
+            if num_str.count(num_str[i]) == 2:
+                return True
```
Comment (laraasphh) [inline Exercises/Justus Exercises/Justus Exercises_03.py:112]: Great solution, I used a lot more lines for this

File: Exercises/Justus Exercises/Justus Exercises_03.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+test_deck(french_deck, 52, "2 of diamonds", "Ace of clubs", "French deck")
+
+# Test Skat Deck
+skat_deck = SkatDeck()
+test_deck(skat_deck, 32, "7 of diamonds", "Ace of clubs", "Skat deck")
+
+print("All deck tests passed!")
+print('\n')
+
+
```
Comment (laraasphh) [inline Exercises/Justus Exercises/Justus Exercises_03.py:94]: I think it's great that you created an extra function to test your classes

File: Exercises/Justus Exercises/Justus Exercises_03.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    def __init__(self, rank, suite):
+        self.rank = rank
+        self.suite = suite
+
+    def __str__(self):
+        return f"{self.rank} of {self.suite}"
+
+    def __repr__(self):
+        return f"Card('{self.rank}', '{self.suite}')"
+
```
Comment (laraasphh) [inline Exercises/Justus Exercises/Justus Exercises_03.py:42]: I don't quite understand why you need the extra class for cards, I've seen it on several people. I only have the FrenchDeck and Skat classes

**Review by laraasphh — APPROVED** [review]
I didn't notice anything negative. I like your approach, keep it up!

**Review by Adrian410 — APPROVED** [review]
I like how you kept your code clean and simple.

File: Exercises/Justus Exercises/Justus Exercises_03.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+def test_deck(deck, expected_length, first_card, last_card, deck_name):
+    # Test length
+    assert len(deck) == expected_length, f"{deck_name} should have {expected_length} cards"
+
+    # Test indexing
+    assert str(deck[0]) == first_card, f"First card should be {first_card}"
+    print(first_card)
+    assert str(deck[-1]) == last_card, f"Last card should be {last_card}"
+    print(last_card)
+    print('\n')
```
[the remark this answers is not in the document]
Comment (Saiki77) [reply Exercises/Justus Exercises/Justus Exercises_03.py]: You are right, this was left over from me testing if the assert test worked as expected. Thank you

**Comment by PixelWelt** [comment]
please merge

**Pull request merged**